packages/v3-sdk/src/functions/swap.ts

```diff
@@ -0,0 +1,238 @@
+        // if we're moving leftward, we interpret liquidityNet as the opposite sign
+        // safe because liquidityNet cannot be type(int128).min
+        if (zeroForOne) liquidityNet *= NEGATIVE_ONE
+
+        state.liquidity = LiquidityMath.addDelta(state.liquidity, liquidityNet)
+      }
+
+      state.tick = zeroForOne ? step.tickNext - 1 : step.tickNext
+    } else if (state.sqrtPriceX96 !== step.sqrtPriceStartX96) {
+      // updated comparison function
+      // recompute unless we're on a lower tick boundary (i.e. already transitioned ticks), and haven't moved
+      state.tick = TickMath.getTickAtSqrtRatio(state.sqrtPriceX96)
+    }
+  }
+
+  return {
+    amountSpecifiedRemaining: state.amountSpecifiedRemaining,
+    amountCalculated: state.amountCalculated,
+    sqrtRatioX96: state.sqrtPriceX96,
+    liquidity: state.liquidity,
+    tickCurrent: state.tick,
+  }
+}
+
+/**
+ * Given an input amount of a token, return the computed output amount, and a pool with state updated after the trade
+ *
+ */
+export const getOutputAmount = async (
+  pool: PoolState,
+  inputAmount: CurrencyAmount<Token>,
+  sqrtPriceLimitX96?: bigint
+): Promise<[CurrencyAmount<Token>, PoolState]> => {
+  invariant(hasInvolvedToken(pool, inputAmount.currency), 'TOKEN')
+
+  const zeroForOne = inputAmount.currency.equals(pool.token0)
+
+  const {
+    amountCalculated: outputAmount,
+    sqrtRatioX96,
+    liquidity,
+    tickCurrent,
+  } = await swap({
+    pool,
+    zeroForOne,
+    amountSpecified: inputAmount.quotient,
+    sqrtPriceLimitX96,
+  })
+  const outputToken = zeroForOne ? pool.token1 : pool.token0
+
+  return [
+    CurrencyAmount.fromRawAmount(outputToken, outputAmount * NEGATIVE_ONE),
+    {
+      ...pool,
+      sqrtRatioX96,
+      liquidity,
+      tick: tickCurrent,
+    },
+  ]
+}
+
+/**
+ * Given a desired output amount of a token, return the computed input amount and a pool with state updated after the trade
+ */
+export const getInputAmount = async (
+  pool: PoolState,
+  outputAmount: CurrencyAmount<Token>,
+  sqrtPriceLimitX96?: bigint
+): Promise<[CurrencyAmount<Token>, PoolState]> => {
+  invariant(outputAmount.currency.isToken && hasInvolvedToken(pool, outputAmount.currency), 'TOKEN')
+
+  const zeroForOne = outputAmount.currency.equals(pool.token1)
+
+  const {
+    amountCalculated: inputAmount,
+    sqrtRatioX96,
+    liquidity,
+    tickCurrent,
+  } = await swap({
+    pool,
+    zeroForOne,
+    amountSpecified: outputAmount.quotient * NEGATIVE_ONE,
+    sqrtPriceLimitX96,
+  })
+
+  const inputToken = zeroForOne ? pool.token0 : pool.token1
+
+  return [
+    CurrencyAmount.fromRawAmount(inputToken, inputAmount),
+    {
+      ...pool,
+      sqrtRatioX96,
+      liquidity,
+      tick: tickCurrent,
+    },
+  ]
+}
+
+/**
+ * Given a desired output amount of a token, return the computed input amount and a pool with state updated after the trade
+ */
+export const getInputAmountByExactOut = async (
```

Comment:
Previously I added this interface to not break the existing ones. Since we're building a brand new pure function now, I think we don't need to follow the exact design. Here's one possible interface I'd suggest

```typescript
const getInputAmount = async (
  pool: PoolState,
  outputAmount: CurrencyAmount<Currency>,
  options?: {
    sqrtPriceLimitX96?: bigint;
    // If exact set to true, throw when insufficient liquidity to get the desired output amount
    exact?: boolean;
  }
)
```